### ISSUE TITLE
FirebaseException should expose code as a property

### ISSUE THREAD

**Issue by rocketraman** [GitLiveApp/firebase-kotlin-sdk#298]
Currently `FirebaseException` can be initialized with a `code` value, however the `code` is only used to build the default Exception message.

There are many cases in which the code is necessary to take the appropriate action -- to take one simple example: https://firebase.google.com/docs/reference/android/com/google/firebase/auth/FirebaseAuthInvalidUserException.

However, `code` is not a property.

Please change

```
FirebaseException(code: String?, cause: Throwable) : Exception("$code: ${cause.message}", cause)
```

to

```
FirebaseException(val code: String?, cause: Throwable) : Exception("$code: ${cause.message}", cause)
```

**Comment by anqus** [comment]
Bump. The only way I've found to handle different errors of the same type is to check what the message contains, as seen below. This is because the messages are formatted differently on iOS and Android, as mentioned in #317: 

```
private fun mapFirebaseAuthErrors(e: FirebaseAuthException): AuthenticationError {
    println("Auth Error: ${e.message}")
    return with(e.message) {
      when {
        this?.contains("There is no user record corresponding to this identifier. The user may have been deleted.") == true -> EmailDoesNotExistError()
        this?.contains("The email address is badly formatted.") == true -> EmailInvalidError()
        this?.contains("The password is invalid or the user does not have a password.") == true -> PasswordWrongError()
        this?.contains("The user account has been disabled by an administrator.") == true -> UserDisabledError()
        this?.contains("The email address is already in use by another account.") == true -> EmailAlreadyExistsError()
        this?.contains("The password must be 6 characters long or more.") == true -> PasswordWeakError()
        else -> AuthenticationError(e.message)
      }
    }
  }
```

It would be much cleaner if we could use the error codes across all platforms that are specified here:

https://github.com/firebase/firebase-ios-sdk/blob/0b3bf6124f6f9643b783c9e74aa9d38d7588b1ff/FirebaseAuth/Sources/Public/FirebaseAuth/FIRAuthErrors.h

**Comment by pmodernme** [comment]
Exposing codes would make our code more robust. Matching message strings is a really flimsy way to handle code, especially when exception handling needs to do more than just present an error to the user.

**Comment by ubuntudroid** [comment]
We don't just need the error code, but also the `details` property that one can pass from Firebase functions and which should end up in `FirebaseFunctionsException` as it does in the official SDK.

**Comment by nbransby** [comment]
> Currently `FirebaseException` can be initialized with a `code` value, however the `code` is only used to build the default Exception message.

`FirebaseException` doesn't have any properties on Android:
https://developers.google.com/android/reference/com/google/firebase/FirebaseException

Did you mean [`FirebaseAuthException`](https://firebase.google.com/docs/reference/android/com/google/firebase/auth/FirebaseAuthException)?

**Comment by nbransby** [comment]
> Did you mean [`FirebaseAuthException`](https://firebase.google.com/docs/reference/android/com/google/firebase/auth/FirebaseAuthException)?

Actually I suspect you were referring to the `FirebaseException` in the JS target as that does have a code that's not exposed  to match https://firebase.google.com/docs/reference/node/firebase.FirebaseError

Exposing it won't help much though as it will only be available to JS sourcesets - I'll see if we can expose the codes the same way the android sdk does and then attempt to map the js and ios code to the android ones

**Comment by ubuntudroid** [comment]
@nbransby I was referring to the Firebase iOS SDK's`Firebase(Functions)Exception`.

**Comment by rocketraman** [comment]
> Actually I suspect you were referring to the `FirebaseException` in the JS target as that does have a code that's not exposed to match https://firebase.google.com/docs/reference/node/firebase.FirebaseError

Yes, sorry I should have specified. I was indeed referring to `FirebaseException` in the JS target:

https://github.com/GitLiveApp/firebase-kotlin-sdk/blob/a28c34ec866a034b6d7bbb92a679b3885d078821/firebase-app/src/jsMain/kotlin/dev/gitlive/firebase/firebase.kt#L55

If the `code` here was made a `val`, then it could be accessed by other code in the JS source sets. It would be great if `code` was in the common source set, but since not all Firebase platforms support/use it, so be it.

> Note the pattern in in Firebas Auth is to use different FirebaseAuthException subtypes for each error type so you can switch on the type instead of code for auth and this is already implemented in the SDK

That's true but not always sufficient. For example, look at the docs for https://firebase.google.com/docs/reference/android/com/google/firebase/auth/FirebaseAuthInvalidUserException.

In order to know *why* the user was invalid, the Firebase team expects you to look at the code. There is no subtype of `FirebaseAuthInvalidUserException` that differentiates these, as far as I know.

**Comment by leoull** [comment]
This is how extract the message for now:

```kt
/**
 * convert FirebaseAuth exception to user friendly message
 * the following error codes and messages are retrieved from [FIRAuthErrorUtils.m]
 * Note: there could be a cleaner way of doing this once [FirebaseException exposes error code as a property](https://github.com/GitLiveApp/firebase-kotlin-sdk/issues/298) is addressed
 */
private fun getErrorDescription(e: Exception): String {
    val message = e.message ?: return "Error occurred, please try again"

    // try extracting user-friendly message
    extractErrorMessage(message)?.let {
        return it
    }

    // there are about 88 error codes, but only the major user errors are parsed here [for example, other types of errors occur because of developer mistakes]
    return when {
        message.contains("ERROR_INVALID_EMAIL") -> "The email address is badly formatted"
        message.contains("ERROR_INVALID_CREDENTIAL") -> "The supplied auth credential is malformed or has expired"
        message.contains("ERROR_USER_DISABLED") -> "The user account has been disabled by an administrator"
        message.contains("ERROR_EMAIL_ALREADY_IN_USE") -> "The email address is already in use by another account"
        message.contains("ERROR_WRONG_PASSWORD") -> "The password is invalid or the user does not have a password"
        message.contains("ERROR_TOO_MANY_REQUESTS") -> "We have blocked all requests from this device due to unusual activity. Try again later."
        message.contains("ERROR_ACCOUNT_EXISTS_WITH_DIFFERENT_CREDENTIAL") -> "An account already exists with the same email address but different sign-in credentials. Sign in using a provider associated with this email address"
        message.contains("ERROR_REQUIRES_RECENT_LOGIN") -> "This operation is sensitive and requires recent authentication. Log in again before retrying this request."
        message.contains("ERROR_PROVIDER_ALREADY_LINKED") -> "User can only be linked to one identity for the given provider"
        message.contains("ERROR_NO_SUCH_PROVIDER") -> "User was not linked to an account with the given provider"
        message.contains("ERROR_NETWORK_REQUEST_FAILED") -> "Network Error. please try again"
        message.contains("ERROR_USER_NOT_FOUND") -> "There is no user record corresponding to this identifier. your account may have been deleted"
        message.contains("ERROR_USER_MISMATCH") -> "The supplied credentials do not correspond to the previously signed in user."
        message.contains("ERROR_WEAK_PASSWORD") -> "The password must be 6 characters long or more"
        message.contains("ERROR_INVALID_VERIFICATION_CODE") -> "invalid verification code"
        else -> "Error occurred, please try again"
    }
}

/**
 * Extract error message.
 *
 * for example, given:
 *      Error Domain=FIRAuthErrorDomain Code=17025
 *      "This credential is already associated with a different user account."
 *      UserInfo={FIRAuthErrorUserInfoNameKey=ERROR_CREDENTIAL_ALREADY_IN_USE,
 *      FIRAuthErrorUserInfoEmailKey=leoul.emiru@gmail.com,
 *      FIRAuthErrorUserInfoUpdatedCredentialKey=<FIROAuthCredential: 0×280152cb0>,
 *      NSLocalizedDescription=This credential is already associated with a different user account.}
 *
 * returns: This credential is already associated with a different user account
 */
private fun extractErrorMessage(errorString: String): String? {
    val regex = "NSLocalizedDescription=(.*?)[.]}".toRegex()
    val matchResult = regex.find(errorString)

    return matchResult?.groupValues?.get(1)
}

/**
 * convert the given exception into another with user-friendly description
 */
fun Exception.makeUserFriendly() = Exception(getErrorDescription(this))
```

**Comment by anggrayudi** [comment]
Status check on this one, since it's been a while and it's been partly overtaken by other work.

**Already done since this was filed** — all cross-platform, declared in `commonMain`:

- `FirebaseFirestoreException.code: FirestoreExceptionCode`
- `FirebaseFunctionsException.code: FunctionsExceptionCode`
- `FirebaseFunctionsException.details: Any?` — which is exactly what @ubuntudroid asked for in this thread

So for Firestore and Functions the pattern @nbransby described (expose codes as a common-source-set enum rather than a raw platform code) is implemented and shipped.

**Still not done — two distinct things:**

1. **The literal request.** `firebase-app/src/jsMain/.../firebase.kt` is unchanged:

   ```kotlin
   public actual open class FirebaseException(code: String?, cause: Throwable) : Exception("$code: ${cause.message}", cause)
   ```

   `code` is still a constructor parameter, not a `val`.

2. **`FirebaseAuthException` has no `code` at all** — only the subtypes. This is the one that actually hurts, and it's why @anqus and @leoull both ended up matching on message strings. As @rocketraman pointed out, subtypes aren't sufficient on their own: `FirebaseAuthInvalidUserException` covers several distinct reasons that Firebase expects you to tell apart *by code*, and there's no subtype split for them. Message matching is also fragile precisely because the message formatting differs per platform (#317).

**The open design question**, which is presumably why this stalled: doing for Auth what was done for Firestore/Functions means picking a shared code space and mapping three different ones onto it — Android's `ERROR_*` strings, iOS's numeric `17xxx` (~88 of them), and JS's `auth/...` strings. That's a real mapping table and an API decision, not a mechanical change.

@nbransby is that still the direction you'd want — an `AuthExceptionCode` enum in `commonMain` with per-platform `actual` mappings, mirroring `FirestoreExceptionCode`/`FunctionsExceptionCode`? If the shape is agreed I'm happy to put a PR together for it. I held off on sending one unprompted precisely because the mapping is opinionated and you'd noted the JS-only `val code` change wouldn't buy much on its own.

**Comment by anggrayudi** [comment]
Correction to my comment above: I said `FirebaseAuthException` has no `code` property. That's wrong — it does, and has since **v2.5.0**. My grep pattern missed the declaration because it's an extension property (`val FirebaseAuthException.code`, not `val code`). Apologies for the noise.

What actually exists, added in #823 (`935c46c5`, 15 Jul 2026):

```kotlin
// commonMain
/** Platform Firebase Auth error code, when the underlying SDK provides one. */
public expect val FirebaseAuthException.code: String?
```

with actuals on every target:

| target | implementation | example value |
|---|---|---|
| Android | `errorCode` | `ERROR_INVALID_EMAIL` |
| JVM | `errorCode` | `ERROR_INVALID_EMAIL` |
| JS | `authCode` | `auth/invalid-email` |
| Apple | `authCode`, set from `code.toString()` | `17008` |

So the headline request in this issue — expose the code as a property — **is done**, and the message-matching workarounds posted earlier in this thread are no longer necessary.

What isn't done is the part @nbransby described as the goal: *"expose the codes the same way the android sdk does and then attempt to map the js and ios code to the android ones."* The property exists on all four targets, but the **values are still three different code spaces**, so `e.code` can't be switched on portably — an Android build sees `ERROR_INVALID_EMAIL` where an iOS build sees `17008`.

Worth noting for Apple specifically: the `ERROR_*` name is already present in the `NSError`, as `userInfo[FIRAuthErrorUserInfoNameKey]` — you can see it in the dumps pasted in #492 and #655 (`FIRAuthErrorUserInfoNameKey=ERROR_CREDENTIAL_ALREADY_IN_USE`, `...=ERROR_INTERNAL_ERROR`). So Apple could report the same strings Android does by reading that key instead of stringifying the numeric code, which would harmonise three of the four targets without inventing a new enum.

That seems a smaller and less disruptive step than the `AuthExceptionCode` enum I asked about earlier. @nbransby @Daeda88 — happy to put up a PR for it if you want that direction; it also unblocks #492, which I've commented on separately.
